pallets/asset-index/src/lib.rs

```diff
@@ -215,19 +235,108 @@ pub mod pallet {
                 Error::<T>::MinimumRedemption
             );
 
-            let deposit = Self::index_token_balance(&caller);
-            ensure!(deposit >= amount, Error::<T>::InsufficientDeposit);
+            let free_balance = T::IndexToken::free_balance(&caller);
+            T::IndexToken::ensure_can_withdraw(
+                &caller,
+                amount,
+                WithdrawReasons::TRANSFER,
+                free_balance.saturating_sub(amount),
+            )?;
 
             let fee = T::WithdrawalFee::withdrawal_fee(amount);
-            let _redeem = amount
+            let redeem = amount
```

Comment:
I have a question:

```text
if an account wants to redeem 10 unit balance exactly && crazily,
how could (s)he make that?
```

I think maybe we should think more about this `sub(&fee)` directly in the amount? Or just leave this question to the client of PINT

Reply:
ideally the user would receive the distribution of the underlying (liquid) assets in proportion to their representation in the index.
Therefor we need to calculate the asset's proportion first and then its PINT equivalent:

```
NAV(amount) = NAV_asset_proportional_0 + NAV_asset_proportional_1 ... + NAV_asset_proportional_n - Fee
```

where `NAV_asset_proportional_n = NAV_asset_n / NAV_liquid * (amount- fee) * Price(PINT/n)`